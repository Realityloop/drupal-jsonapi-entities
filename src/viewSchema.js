import schemaBase from './schemaBase'

import entityViewDisplay from './resources/entityViewDisplay'
import fieldConfig from './resources/fieldConfig'

class viewSchema extends schemaBase {
  constructor(entityType, bundle, mode, context) {
    super(entityType, bundle, mode, context)

    this.resources = [
      new entityViewDisplay(this),
      new fieldConfig(this)
    ]
  }

  processFields() {
    const fields = {}

    for (const resource of this.resources) {
      for (const field in resource.fields) {
        if (typeof fields[field] === 'undefined' && resource instanceof entityViewDisplay) {
          fields[field] = resource.fields[field]
        }

        else if (typeof fields[field] !== 'undefined') {
          // Remove settings, they're merged seperately.
          const settings = resource.fields[field].settings
          delete resource.fields[field].settings

          fields[field] = Object.assign(fields[field], resource.fields[field])
          fields[field].settings = Object.assign(fields[field].settings, settings)
        }
      }
    }

    return fields
  }

  processGroups() {
    return this.resources[0].groups
  }
}

export default viewSchema
