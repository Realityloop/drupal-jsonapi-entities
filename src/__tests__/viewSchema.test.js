import viewSchema from '../viewSchema'

import mockEntityViewDisplay from '../resources/__fixtures__/entity_view_display'
import mockFieldConfig from '../resources/__fixtures__/field_config'

const entityType = 'node'
const bundle = 'recipe'
const mode = 'default'

const mockAxios = {
  post: () => ({
    data: {
      'node--recipe--default--entity_view_display': { body: JSON.stringify(mockEntityViewDisplay) },
      'node--recipe--field_config': { body: JSON.stringify(mockFieldConfig) }
    }
  })
}

test('viewSchema', async () => {
  const schema = new viewSchema({ entityType, bundle, mode, axios: mockAxios })
  await schema.build()

  expect(schema.fields.items).toMatchSnapshot()
  expect(schema.keys).toMatchSnapshot()
  expect(schema.fields.get()).toMatchSnapshot()
  expect(schema.fields.get(false)).toMatchSnapshot()
})