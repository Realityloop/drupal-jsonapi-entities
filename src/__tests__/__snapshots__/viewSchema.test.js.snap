// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewSchema 1`] = `
Object {
  "field_cooking_time": Object {
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 1,
  },
  "field_difficulty": Object {
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "list_default",
    "weight": 2,
  },
  "field_image": Object {
    "description": "",
    "id": "field_image",
    "label": "Image",
    "labelPosition": "hidden",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "image_link": "",
      "responsive_image_style": "3_2_image",
    },
    "thirdPartySettings": Array [],
    "type": "responsive_image",
    "weight": 3,
  },
  "field_ingredients": Object {
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": 4,
  },
  "field_number_of_servings": Object {
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": "",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 5,
  },
  "field_preparation_time": Object {
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 6,
  },
  "field_recipe_category": Object {
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 7,
  },
  "field_recipe_instruction": Object {
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 8,
  },
  "field_summary": Object {
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "labelPosition": "hidden",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 9,
  },
  "field_tags": Object {
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 10,
  },
  "links": Object {
    "id": "links",
    "labelPosition": undefined,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": undefined,
    "weight": 0,
  },
  "title": Object {
    "id": "title",
    "labelPosition": "hidden",
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": -5,
  },
}
`;

exports[`viewSchema 2`] = `
Array [
  "field_cooking_time",
  "field_difficulty",
  "field_image",
  "field_ingredients",
  "field_number_of_servings",
  "field_preparation_time",
  "field_recipe_category",
  "field_recipe_instruction",
  "field_summary",
  "field_tags",
  "links",
  "title",
]
`;

exports[`viewSchema 3`] = `
Array [
  Object {
    "id": "title",
    "labelPosition": "hidden",
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": -5,
  },
  Object {
    "id": "links",
    "labelPosition": undefined,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": undefined,
    "weight": 0,
  },
  Object {
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 1,
  },
  Object {
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "list_default",
    "weight": 2,
  },
  Object {
    "description": "",
    "id": "field_image",
    "label": "Image",
    "labelPosition": "hidden",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "image_link": "",
      "responsive_image_style": "3_2_image",
    },
    "thirdPartySettings": Array [],
    "type": "responsive_image",
    "weight": 3,
  },
  Object {
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": 4,
  },
  Object {
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": "",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 5,
  },
  Object {
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 6,
  },
  Object {
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 7,
  },
  Object {
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 8,
  },
  Object {
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "labelPosition": "hidden",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 9,
  },
  Object {
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 10,
  },
]
`;

exports[`viewSchema 4`] = `
Array [
  Object {
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 1,
  },
  Object {
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "list_default",
    "weight": 2,
  },
  Object {
    "description": "",
    "id": "field_image",
    "label": "Image",
    "labelPosition": "hidden",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "image_link": "",
      "responsive_image_style": "3_2_image",
    },
    "thirdPartySettings": Array [],
    "type": "responsive_image",
    "weight": 3,
  },
  Object {
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": 4,
  },
  Object {
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": "",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 5,
  },
  Object {
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "prefix": "",
      "prefix_suffix": true,
      "suffix": " minutes",
      "thousand_separator": "",
    },
    "thirdPartySettings": Array [],
    "type": "number_integer",
    "weight": 6,
  },
  Object {
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 7,
  },
  Object {
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "labelPosition": "above",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 8,
  },
  Object {
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "labelPosition": "hidden",
    "property": false,
    "required": true,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": "text_default",
    "weight": 9,
  },
  Object {
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "labelPosition": "above",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "link": true,
    },
    "thirdPartySettings": Array [],
    "type": "entity_reference_label",
    "weight": 10,
  },
  Object {
    "id": "links",
    "labelPosition": undefined,
    "settings": Object {},
    "thirdPartySettings": Array [],
    "type": undefined,
    "weight": 0,
  },
  Object {
    "id": "title",
    "labelPosition": "hidden",
    "settings": Object {
      "link_to_entity": false,
    },
    "thirdPartySettings": Array [],
    "type": "string",
    "weight": -5,
  },
]
`;
